rfcs/53-consistent-hierarchy.md

```diff
@@ -0,0 +1,410 @@
+expensive to run, so the documentation for them should reflect this and
+discourage misuing it.
+
+### HierarchyEvents
+Three new events should be added here, each corresponding to a specific change
+in the hierarchy.
+
+```rust
+pub enum ChildAdded {
+  pub parent: Entity,
+  pub child: Entity,
+}
+
+pub enum ChildRemoved {
+  pub parent: Entity,
+  pub child: Entity,
+}
+
+pub enum ChildMoved {
+  pub parent: Entity,
+  pub previous_parent: Entity,
+  pub new_parent: Entity,
+}
+```
+
+Both `ChildAdded` and `ChildRemoved` will be fired when a child moved occurs.
+These can be used to detect changes where `PreviousParent` is currently used.
+
+### `bevy_hierarchy`-`bevy_transform` split.
+These hierarchy types should be moved to a new `bevy_hierarchy` crate to avoid
+requiring compiling the hierachy alongside the default "f32-based affine
+transform system in Euclidean space". This should allow users to create their own
+transform systems (i.e. f64-based transforms, fixed-point transforms,
+non-Euclidean spaces) while also being able to leverage the same
+parent-sibling-child hierarchies currently found in `bevy_transform`. This should
+also allow us to internally leverage the same decoupling too for specialized
+transforms for 2D and UI.
+
+## Benefits
+
+ - This rocks the boat the least compared to the following alternatives and
+   largely provides the same benefits.
+ - This design has (nearly) guarenteed global consistency. The hierarchy cannot
+   be arbitrarily mutated in inconsistent between synchronization points.
+   Changes are deferred to a until the end of the current stage instead of a
+   single global maintainence system.
+ - The cache locality when iterating over the immediate children of an entity is
+   retained. This saves an extra cache miss per entity when traversing the
+   hierarchy.
+ - The removal of `PreviousParent` should reduce the archetype fragmentation of
+   the current solution.
+ - Any update to the hierarchy is "atomic" and is `O(n)` in the number of
+   children an entity has. For any practical hierarchy, this will typically be
+   a very small n.
+ - Existing query filters still work as is. For example, `Without<Parent>` can
+   still be used to find root entities, and `Changed<Children>` can be used to
+   find any entity with either new children, removed children, or reordered
+   children.
+ - The hierarchy is decoupled from the base transform system, allowing both Bevy
+   developer and third-party crates to make their own independent transform
+   system without compiling the default transform system.
+
+## Drawbacks
+
+ - The fact that the hierarchy is not optimally stored in memory is still an
+   issue, and all hiearchy traversals require heavy random access into memory.
+ - Updates are still not immediately visible within a stage and deferred until
+   the next command buffer flush.
+ - Hierarchy updates are now single threaded.
```

Comment:
Not 100% correct, the existing Hierarchy update system already only used a single thread, though it could be run in parallel with other systems.

Reply:
Haha! Here is definitely where the current inconsistencies come in. Maintenance is single threaded, mutation could be parallelized, by using disjoint queries for Parent in multiple systems. I'll update the text to reflect this.